### DISCUSSION TITLE
Allow to specify which files are scripts and which are modules

### DISCUSSION THREAD

**Discussion by christophehenry** [biomejs/biome#4207] [Ideas]
The `useStrictMode`, currently in nursery, allows to enforce the use of the directive "use strict" in script files. However, documentation states:

> Biome considers a CommonJS (.cjs) file as a script file. By default, Biome recognizes a JavaScript file (.js) as a module file, except if "type": "commonjs" is specified in `package.json`.

Which is not a convenient way for detecting which files are scripts and which are modules. It requires either to use a `package.json` which is inconvenient for project that don't use NodeJS or rename every script file in the project to use the `.cjs` extension, which is also inconvenient. Some project still use the old-school HTML of sourcing files (`<script src="…">`) and use neither Node or CommonJS.

I believe there should be configuration in `biome.json`, (like [`eslint` does](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects)) possibly in the [`files`](https://biomejs.dev/reference/configuration/#files) section and it should also be configurable per-file using a comment.

**Comment by ematipico** [comment]
Creating **yet another option** is a rabbit hole and source of many inconsistencies across projects.

I prefer to make a rule smarter and collect as many metadata as possible from the environment, instead adding options. An option wouldn't make the rule portable in this case, and it forces the developer to babysit Biome. Biome wants to make the DX easier, not the other way around. 

If we need to evaluate an option, I would prefer another part of the configuration.

Going back to the proposal, I think you're onto something, and checking the manifest can't be the only way to deduce the file type.

Are there other cases we would need to be aware of?

**Reply by arendjr** [reply]
I'm not fond of the idea of configuring it per file either, but I think having a setting that defines the source type for `.js` might be unavoidable. Of course when our HTML support is ready, we could scan HTML files for `<script>` tags and treat those appropriately, but it would still not satisfy all use cases: scripts loaded dynamically by adding a `<script>` tag to the DOM wouldn't be detected, for instance. And situations in which the HTML is generated by a backend wouldn't be detectable for us either.

Looking at the `package.json` is a good heuristic, but having an option of the same effect in our own configuration wouldn't make things much more complicated, I think.

**Reply by christophehenry** [reply]
> Are there other cases we would need to be aware of?

I'm currently trying to add Biome as a linter to a farily old-fashion Django project: no ties to Node (and none desired), some JS files are modules, some are scripts and nothing in the context specifies which is which except how they're sourced in the template. So it would be quite difficult to deduce much from the envrironement.

Note that make script to use the `.cjs` extension would work but wouldn't make much sense since they're actually not CommonJS files. And I'd like to enforce the `"use strict"` on scripts.

**Reply by christophehenry** [reply]
However, having a configuration to say "every `.js` file here is a script" and having Biome to automatically detect `.mjs` files as modules would work for me.
